smolyay/surrogate.py

```diff
@@ -1,546 +1,742 @@
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
         if oob:
-            raise ValueError('x must lie in domain of surrogate')
-        # transform point into basis domain and evaluate
-        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
-        value = 0
-        for coeff, basis in zip(self._coefficients, self.grid.basis_functions):
-            if self.dimension > 1:
-                term = numpy.prod(
-                        [basis[i](x_scaled[...,i]) for i in range(len(basis))],
-                        axis=0)
-            else:
-                term = basis(x_scaled)
-            value += coeff*term
+            raise ValueError("X must lie in domain of surrogate.")
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = [
+            self.basis_sets[dim](X[:, dim], self.domain[dim])
+            for dim in range(self.num_dimensions)
+        ]
+
+        # use lookup table to combine terms
+        answer = numpy.ones(len(X)) * self._integration_constant
+        for ic, coeff in zip(self._index_combinations, self._coefficients):
+            answer = answer + numpy.real(
+                coeff
+                * numpy.prod(
+                    [lookup_table[dim][ic[dim], :] for dim in range(len(ic))], axis=0
+                )
+            )
+
+        # return results
+        if len(X) == 1:
+            return answer[0]
+        else:
+            return answer
+
+    def predict_gradient(self, X):
+        """Evaluate gradient or Jacobian of the surrogate at a given input.
 
-        if (self.dimension == 1 and input_shape == ()) or (self.dimension > 1 and len(input_shape) == 1):
-            value = value.item()
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated
+
+        Returns
+        -------
+        ndarray of shape (n_samples,) or (n_samples, num_dimensions)
+            Gradient output at x.
+
+        Raises
+        ------
+        RuntimeError
+            For surrogate to be evaluated, function needs to be fit.
+        IndexError
+            Input must be 2D array with shape (n_samples, n_features).
+        ValueError
+            Input must lie in domain of surrogate.
+        """
+        # validate inputs
+        if not self._valid_cache:
+            raise RuntimeError("Model must be fit!")
+        X = numpy.array(X, ndmin=2)
+        if X.shape[1] != self.num_dimensions:
+            raise IndexError("Must be 2D array with shape (n_samples, n_features).")
+        oob = any(
+            numpy.any(X[:, i] < self.domain[i][0])
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
+        if oob:
+            raise ValueError("X must lie in domain of surrogate.")
```

Comment:
This code is getting reused, should we make a private method that can do this?

Reply:
Yes.